modules/apps/journal/journal-service/src/main/java/com/liferay/journal/internal/transformer/DLPreviewURLRefreshTimestampTransformerListener.java

```diff
@@ -0,0 +1,155 @@
+ * details.
+ */
+
+package com.liferay.journal.internal.transformer;
+
+import com.liferay.document.library.kernel.exception.NoSuchFileEntryException;
+import com.liferay.document.library.kernel.service.DLAppLocalService;
+import com.liferay.journal.constants.JournalPortletKeys;
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.repository.model.FileEntry;
+import com.liferay.portal.kernel.templateparser.BaseTransformerListener;
+import com.liferay.portal.kernel.templateparser.TransformerListener;
+import com.liferay.portal.kernel.util.Validator;
+
+import java.util.Date;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author István András Dézsi
+ */
+@Component(
+	immediate = true,
+	property = "javax.portlet.name=" + JournalPortletKeys.JOURNAL,
+	service = TransformerListener.class
+)
+public class DLPreviewURLRefreshTimestampTransformerListener
+	extends BaseTransformerListener {
+
+	@Override
+	public String onOutput(
+		String output, String languageId, Map<String, String> tokens) {
+
+		if (_log.isDebugEnabled()) {
+			_log.debug("onOutput");
+		}
+
+		return replace(output);
+	}
+
+	protected FileEntry getFileEntry(Matcher matcher) throws PortalException {
+		if (matcher.group(5) != null) {
+			long groupId = Long.valueOf(matcher.group(2));
+
+			String uuid = matcher.group(5);
+
+			return _dlAppLocalService.getFileEntryByUuidAndGroupId(
+				uuid, groupId);
+		}
+
+		long groupId = Long.valueOf(matcher.group(2));
+		long folderId = Long.valueOf(matcher.group(3));
+		String title = matcher.group(4);
+
+		try {
+			return _dlAppLocalService.getFileEntry(groupId, folderId, title);
+		}
+		catch (NoSuchFileEntryException noSuchFileEntryException) {
+			if (_log.isDebugEnabled()) {
+				_log.debug(noSuchFileEntryException, noSuchFileEntryException);
+			}
+
+			return _dlAppLocalService.getFileEntryByFileName(
+				groupId, folderId, title);
+		}
+	}
+
+	protected String getReplacement(Matcher matcher) throws PortalException {
+		FileEntry fileEntry = getFileEntry(matcher);
+
+		if (fileEntry == null) {
+			return matcher.group(0);
+		}
+
+		Date modifiedDate = fileEntry.getModifiedDate();
+
+		long timeStamp = modifiedDate.getTime();
+
+		StringBundler sb = new StringBundler(4);
+
+		sb.append(matcher.group(1));
+		sb.append("?t=");
+		sb.append(timeStamp);
+		sb.append(matcher.group(6));
+
+		return sb.toString();
+	}
+
+	protected String replace(String s) {
+		if (Validator.isNull(s) || !s.contains("<img") ||
+			!s.contains("/documents/")) {
+
+			return s;
+		}
+
+		Matcher matcher = _pattern.matcher(s);
+
+		try {
+			StringBuffer sb = null;
```

Comment:
Can you use StringBuilder here? StringBuffer has some synchronization issues.

Reply:
I intentionally using `StringBuffer` because `matcher.appendReplacement()` doesn't accept` StringBuilder`.
`BaseRegexStringContentTransformer` uses `StringBuffer` for individual group replacement as well, which is quite a tricky maneuver, and I couldn't find a method that can be achieved with the `StringBuilder`. However, I'm using `StringBuilder` elsewhere in `DLPreviewURLRefreshTimestampTransformerListener.java`.

Reply:
I see :facepalm:

Reply:
Sorry, I should have mentioned it in the initial description. I noticed that StringBuffer is deprecated as well.